src/libraries/Common/src/Interop/Unix/System.Native/Interop.ReadLink.cs

```diff
@@ -20,24 +21,32 @@ internal static partial class Sys
         /// Returns the number of bytes placed into the buffer on success; bufferSize if the buffer is too small; and -1 on error.
         /// </returns>
         [DllImport(Libraries.SystemNative, EntryPoint = "SystemNative_ReadLink", SetLastError = true)]
-        private static extern int ReadLink(string path, byte[] buffer, int bufferSize);
+        private static extern int ReadLink(ref byte path, byte[] buffer, int bufferSize);
 
         /// <summary>
         /// Takes a path to a symbolic link and returns the link target path.
         /// </summary>
-        /// <param name="path">The path to the symlink</param>
-        /// <returns>
-        /// Returns the link to the target path on success; and null otherwise.
-        /// </returns>
-        public static string? ReadLink(string path)
+        /// <param name="path">The path to the symlink.</param>
+        /// <returns>Returns the link to the target path on success; and null otherwise.</returns>
+        internal static string? ReadLink(ReadOnlySpan<char> path)
         {
-            int bufferSize = 256;
+            int outputBufferSize = DefaultPathBufferSize;
+
+            // Use an initial buffer size that prevents disposing and renting
+            // a second time when calling ConvertAndTerminateString.
+            int pathBufferSize = Encoding.UTF8.GetMaxByteCount(path.Length) + 1;
+            using var converter = new ValueUtf8Converter(stackalloc byte[pathBufferSize]);
```

Comment:
Is it safe to `stackalloc` `pathBufferSize`?

Reply:
I did this only because the method below, `converter.ConvertAndTerminateString`, is going to do the same because we are going to pass the `path`, so I wanted to avoid the extra step of increasing its size.

Granted, this could be wrong. @GrabYourPitchforks is this `stackalloc` (and the one in `ConvertAndTerminateString`) safe?

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/Common/src/System/Text/ValueUtf8Converter.cs#L24-L38

Reply:
There is no _stackalloc_ in `ConvertAndTerminateString`; it's always using the pool. The problem with allowing arbitrary-sized lengths for _stackalloc_ is that you could overflow the stack. These exceptions are non-recoverable and fatal to the application. If remote user input can trigger such an exception, it's considered a security vulnerability.

The typical way to resolve this is to do what Adam suggests below - use a constant value with stackalloc. Not only is it safer - it's _faster_.